### PR TITLE
Remove `ax` from runtime 

### PR DESCRIPTION
**Pull request by liamqma**
This PR is the first attempt to remove `ax` from runtime. It will remove `ax` if the below conditions are met:
- Use CSS props 
- Does not receive className
- Atomic CSS rules are statically defined

Example
```
<div css={{ color: 'red' }} />
```

### CODE REVIEW

File: packages/babel-plugin/src/utils/build-compiled-component.ts [hunk 1]
```diff
@@ -57,6 +58,49 @@ const getExpression = (
   return value;
 };
 
+/**
+ * Check if 'node' is a 'inBuiltComponent'.
+ * We assume it's a 'inBuiltComponent' if it starts with a lowercase letter (i.e. div).
+ *
+ * @param node Originating node
+ */
+const isInBuiltComponent = (node: t.JSXElement): boolean => {
```
Comment (liamqma) [inline packages/babel-plugin/src/utils/build-compiled-component.ts:67]: - [ ] Add a test case for "UserDefinedComponent"

File: packages/babel-plugin/src/utils/build-compiled-component.ts [hunk 2]
```diff
@@ -57,6 +58,49 @@ const getExpression = (
...
+ * We assume it's required if
+ * - there is a class name prop statically defined
+ * - there is conditional css
+ * - 'node' is a 'UserDefinedComponent'
+ *
+ * @param node Originating node
+ * @param classNames The classNames for a component at runtime
+ *
+ */
+const isAxRequired = (node: t.JSXElement, classNames: t.Expression[]) => {
```
Comment (liamqma) [inline packages/babel-plugin/src/utils/build-compiled-component.ts:97]: - [ ] Add a test case for dynamic atomic declaration

File: packages/babel-plugin/src/utils/build-compiled-component.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import generate from '@babel/generator';
 import template from '@babel/template';
 import * as t from '@babel/types';
+import { ax } from '@compiled/react/runtime';
```
Comment (liamqma) [inline packages/babel-plugin/src/utils/build-compiled-component.ts:4]: Not sure if a babel plugin should import stuff from runtime. I've been thinking about moving `ax` to `@compiled/utils`. But I noticed `@compiled/utils` requires some modules (e.g. source-map) which shouldn't exist in runtime. Although bundler might remove it from runtime code, I think it's still dangerous for runtime to depend on `@compiled/utils`.

It's safe other way around. (build time library depends on runtime library)

File: packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts [hunk 4]
```diff
@@ -182,15 +182,15 @@ describe('css prop behaviour', () => {
       import '@compiled/react';
 
       const base = { color: 'black' };
-      const top = \` color: red; \`;
+      const top = \` marginTop: 10px; \`;
```
Comment (pancaspe87) [inline packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts:185]: Why was this changed?

File: packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts [hunk 5]
```diff
@@ -182,15 +182,15 @@ describe('css prop behaviour', () => {
...
 
       <div css={[base, top]}>hello world</div>
     `);
 
     expect(actual).toIncludeMultiple([
-      '._syaz5scu{color:red}',
       '._syaz11x8{color:black}',
-      '<div className={ax(["_syaz11x8","_syaz5scu"])}>hello world</div>',
```
Comment (pancaspe87) [inline packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts:193]: I never looked into how [composition](https://compiledcssinjs.com/docs/composition#composition) works, but I would have expected only 1 class (`_syaz5scu `) in here in the original example

File: packages/babel-plugin/src/utils/build-compiled-component.ts [hunk 6]
```diff
@@ -57,6 +58,49 @@ const getExpression = (
...
+ * - there is conditional css
+ * - 'node' is a 'UserDefinedComponent'
+ *
+ * @param node Originating node
+ * @param classNames The classNames for a component at runtime
+ *
+ */
+const isAxRequired = (node: t.JSXElement, classNames: t.Expression[]) => {
+  return (
+    isInBuiltComponent(node) &&
```
Comment (pancaspe87) [inline packages/babel-plugin/src/utils/build-compiled-component.ts:99]: maybe I misunderstood but thought that `ax` is not required in cases such as `<div css={css({color: 'red'})} />`

File: packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts (hunk 5 above)
Comment (liamqma) [reply packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts:193]: Yea, only 1 class would exist at runtime after `ax` is executed. Previously the babel-plugin wouldn't do any duplication. It just simply constructs something like `ax(["_syaz11x8","_syaz5scu"]` and leave it to runtime `ax` to ensure one atomic group (color) exists.

File: packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts (hunk 4 above)
Comment (liamqma) [reply packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts:185]: because I created another test case to [test deduping atomic group](https://github.com/atlassian-labs/compiled/pull/1354/files#diff-69558aafb510bb07e99ed2dc72efeb29d5ff2370b0a3910c2a996c71f4081fa9R737). So I updated this one to not have duplicated `color`.

File: packages/babel-plugin/src/utils/build-compiled-component.ts (hunk 6 above)
Comment (liamqma) [reply packages/babel-plugin/src/utils/build-compiled-component.ts:99]: Sorry, I don't quite follow your question. Can you elaborate it please?

File: packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts (hunk 5 above)
Comment (pancaspe87) [reply packages/babel-plugin/src/css-prop/__tests__/behaviour.test.ts:193]: yeahhh, makes sense.

I think I was getting confused by the Styled API behaviour; IIRC it would generate only one class for each property

File: packages/babel-plugin/src/utils/build-compiled-component.ts (hunk 6 above)
Comment (pancaspe87) [reply packages/babel-plugin/src/utils/build-compiled-component.ts:99]: nvm, I just read the description... I get it now :)

File: packages/babel-plugin/src/utils/build-compiled-component.ts (hunk 3 above)
Comment (JakeLane) [reply packages/babel-plugin/src/utils/build-compiled-component.ts:4]: I think this is ok as we don't import from the babel plugin in the react package. 

As for side-effects in the runtime accidentally being imported, we should be safe if we use the `size-limit` tool to prevent regressions.

File: packages/babel-plugin/src/utils/build-compiled-component.ts [hunk 7]
```diff
@@ -57,6 +58,49 @@ const getExpression = (
...
+  ) {
+    tagName = node.openingElement.name.name;
+  } else {
+    tagName = node.openingElement.name.property.name;
+  }
+
+  if (typeof tagName !== 'string') return false;
+
+  // Check if the first letter is lower case
+  return tagName.charCodeAt(0) >= 97 && tagName.charCodeAt(0) <= 122;
```
Comment (JakeLane) [inline packages/babel-plugin/src/utils/build-compiled-component.ts:83]: I don't think this is safe as we could have a user-defined component with a lower case name. We should find an alternative way to decide if this is a dom component.

File: packages/babel-plugin/src/utils/build-compiled-component.ts [hunk 8]
```diff
@@ -57,6 +58,49 @@ const getExpression = (
...
+};
+
+/**
+ * Check if the `ax` runtime utility is required.
+ * We assume it's required if
+ * - there is a class name prop statically defined
+ * - there is conditional css
+ * - 'node' is a 'UserDefinedComponent'
```
Comment (JakeLane) [inline packages/babel-plugin/src/utils/build-compiled-component.ts:91]: props spreading is also another way for the component to have a defined classname

**Review by JakeLane — CHANGES_REQUESTED** [review]

File: packages/babel-plugin/src/utils/build-compiled-component.ts (hunk 7 above)
Comment (liamqma) [reply packages/babel-plugin/src/utils/build-compiled-component.ts:83]: Yup, you are right. 👍

File: packages/babel-plugin/src/utils/build-compiled-component.ts (hunk 8 above)
Comment (liamqma) [reply packages/babel-plugin/src/utils/build-compiled-component.ts:91]: Good point! 👍

- [ ] consider props spreading 

I notice props spreading can cause issue in existing/master codebase too. Raised an issue: https://github.com/atlassian-labs/compiled/issues/1361

**Comment by liamqma** [comment]
Close this PR for now because removing `ax` is no longer a prerequisite of compressing class names. [Instead, we update 'ax' to accept short class name](https://github.com/atlassian-labs/compiled/pull/1396). The reason is we can only remove `ax` in limited cases. Even in those cases, it requires AST tree modification, which is prone to bugs. 
We'll revisit removing `ax` as runtime improvement holistically in the future.
